### ISSUE TITLE
[Meta-Ticket] Alignment I/O

### ISSUE THREAD

**Issue by rrahn** [seqan/seqan3#316]
# Concepts
- [x] Aligned Sequence Concept (#352)
- [ ] Alignment Input File Format Concept
- [x] Alignment Output File Format Concept

# Classes
- ~Alignment Class~ update : There is no alignment class anymore since we just have a pair of aligned sequences and a class wrapper would introduce unneccessary overhead.
  - [ ] replace `stream_alignment` function from PR #233 when the concept of a "seqan pretty debug 
stream" is implemented (#121)  
  - [ ] close PR #233 if code is really not needed anymore
  - [x] Add `get_cigar_vector` function (introduced in PR #414)
- [x] SAM tag dictionary (#415)
    - [ ] PR #415 introduced a consexpr string literal in file sam_tag_dictionary.hpp that introduces a warning on clang. If we support clang and require C++20 on GCC we can replace this with a constexpr string. This bullet point is also in meta issue #404 
- [ ] cigar_op and cigar alphabet class (#414)
- [ ] Alignment Input File
- [x] Alignment Output File
- [ ] Dummy Sequence Class (Always throws on access)
# Specific Alignment Formats
- [x] Sam Format
- [ ] Bam Format
- [ ] Blast Format (For time and simplicity reasons only the tabular format is considered for now and the Blast format will only be available as output format)

# Parsing
* Read binary data
* Write binary data
* Endianness
* Formatted Numbers

# Dependencies 
* BGZF input stream (#317)
* BGZF output stream (#317)

**Comment by h-2** [comment]
I agree in general, but I would like to stress that the general design of the alignment i/o is entirely independent of compression which is only required for BAM. 

So I would propose the order:

1. alignment object
2. alignment file [in|out] format concept
3. alignment file [in|out]
4. alignment file format sam

Anything else after that or in parallel. But the above would be sufficient for a minimal working example in the workshop.

**Comment by rrahn** [comment]
I know, I just wanted to drop some suff. It needs some more fine grain ordering. 
Here is my intention: I can't follow all tickets if they are spread over the entire repo. Since we need to port entire modules, I wanted to establish a top-level structure, that helps to easily find issues and establish a better project controlling/monitoring. these Meta-tickets are there to collect general topics. These can then be more fine grained. For every bullet point a new issue is generated, which is referenced here. So you and I can control, what are the next steps to do. In a weekly team meeting these issues should be discussed and assigned to the team.
So please feel free to reorganise if you have more thoughts already.

**Comment by rrahn** [comment]
Of course, these meta-issues should be about the entire module, not just a specific milestone. But for the milestone we define the issues that should be compiled from this meta-ticket.

**Comment by smehringer** [comment]
Summary Alignment Record Discussion
----------------------------------------------------

The key field of the alignment record will be `seqan3::field::alignment` that features an `seqan3::alignment` object with two sequences. The alignment IO thereby only considers pairwise alignment as multiple alignments are a different use case. The two sequences in the alignment object will preferably be of the `gap_decorator` type that only stores the gap information and a reference to the `seqan3::field::SEQ`/`seqan3::field::REF_SEQ` objects. 

For SAM/BAM file the alignment is usually incomplete since the reference sequence is unknown. This will be handled by introducing a dummy/proxy sequence that consists of `N`'s. If the reference characters are queried, an exception is thrown to warn the user that no reference information is available. The alignment file (at least for BAM/SAM format) should be also constructable with a second file, the reference sequence file, in which case the reference is loaded and the alignments properly instantiated. There is also the idea of "lazy loading" where the provided reference sequence file is not opened unless an alignment is queried for reference related information, at which point the reference sequence file is accessed (via the index?) and the corresponding information extracted.

**Comment by smehringer** [comment]
TODO: Right now the REF_ID field for the SAM format is a string but it should be a number, like in BAM files. That way SAM and BAM formats are consistent about the REF_ID

**Comment by smehringer** [comment]
All relevant information have been transferred to the wiki page
